### PR TITLE
Add buttons for Async Clipboard API

### PR DESCRIPTION
**Pull request by garykac**

### CODE REVIEW

**Comment by snffy** [comment]
Thanks for your pull request! Do you mind going to your fork -> Settings -> Repositories -> Github Pages and publish it so I can have a play?

**Comment by garykac** [comment]
Done: https://garykac.github.io/permission.site/

**Comment by garykac** [comment]
Note that the code to hook up the permission prompts to the API has not landed yet. I have a CL pending that hooks them up that should be landing soon.

File: index.html [hunk 1]
```diff
@@ -59,6 +56,17 @@
...
+
+<p>Async Clipboard API</p>
+
+<button id="read-text">Read text</button>
+<button id="write-text">Write text</button>
+<br>
+<button id="read-text-delayed">Read text (delayed)</button>
+<button id="write-text-delayed">Write text (delayed)</button>
+
+<hr>
```
Comment (snffy) [inline index.html]: Please delete this last `<hr>` tag - I think the last row of buttons isn't usually followed by one.

File: index.html [hunk 2]
```diff
@@ -80,6 +88,17 @@
...
+<tr>
+  <td>Async Clipboard</td>
+  <td>
+    Implemented behind the experimental flag <code>chrome://flags/#enable-experimental-web-platform-features</code>.
+    <br>
+    To enable the Content Settings UX for setting clipboard permission, you'll also need to enable
+    <code>chrome://flags/#clipboard-content-setting</code>
+    <br>
+    Note: Only available for secure connections (https).
+  </td>
```
Comment (snffy) [inline index.html]: I'm pretty unfamiliar with the clipboard stuff, so please correct me if I'm wrong/misunderstanding something - can you also explain the difference between the existing button that reads 'Copy' and the new 'Write text' buttons you're adding?

File: index.js [hunk 3]
```diff
@@ -307,7 +307,67 @@ window.addEventListener("load", function() {
       } catch(e) {
         displayOutcome("protocol-handler", "error")(e);
       }
-    }
+    },
+
+    "read-text": function() {
+      console.log("read-text");
```
Comment (snffy) [inline index.js]: To be consistent, I think we should leave these extra logs out (but definitely keep the ones on success/deny).

File: index.js [hunk 4]
```diff
@@ -307,7 +307,67 @@ window.addEventListener("load", function() {
...
+      var cb = navigator.clipboard;
+      if (cb) {
+        cb.readText().then(function(data) {
+          console.log("Successfully read data from clipboard: '" + data + "'");
```
Comment (snffy) [inline index.js]: Please try using the displayOutcome() function? (For this one and all the other logs below)

File: index.js [hunk 5]
```diff
@@ -307,7 +307,67 @@ window.addEventListener("load", function() {
...
+        }, function() {
+          console.log("Failed to read from clipboard");
+        });
+      } else {
+        console.log("navigator.clipboard not available");
+      }
+    },
+
+    "write-text": function() {
+      console.log("write-text");
```
Comment (snffy) [inline index.js]: Please delete this log as well.

**Comment by snffy** [comment]
Ah that makes sense - it was still helpful because of your console messages though. Thanks!

File: index.html (hunk 1 above)
Comment (garykac) [reply index.html]: Done.

File: index.html (hunk 2 above)
Comment (garykac) [reply index.html]: The existing button is for the old execCommand() access to the clipboard which only provides write access and is not protected by any permissions (only a user gesture).

The new buttons are for the new Async Clipboard API (https://w3c.github.io/clipboard-apis/).
The new API provides read/write access and has more protections (Secure context; permission) than the old API.

File: index.js (hunk 3 above)
Comment (garykac) [reply index.js]: Done

File: index.js (hunk 5 above)
Comment (garykac) [reply index.js]: Done.

File: index.html (hunk 2 above)
Comment (snffy) [reply index.html]: Oh, that makes sense. Thanks very much - do you mind also including this in your note in the table here? Maybe something like "Unlike the existing Copy button, this uses the newer [API](https://w3c.github.io/clipboard-apis/)."

Comment (garykac) [reply index.html]: Done. Let me know if you think there should be any additional info in the table.

**Comment by snffy** [comment]
Thanks for your PR :)

**Comment by garykac** [comment]
Thanks!

**Comment by lgarron** [comment]
We still need to migrate the hosting to a Google-owned App Engine project, but I've deployed this change on the old server manually for now.

**Pull request merged**